### PR TITLE
Added exception handling to WaitAll in DeletionService

### PR DESCRIPTION
**Pull request by mikaelweave**
## Description
Added needed exception handling and logging in `DeletionService` for `DeleteMultipleAsync`. 

## Related issues
AB#115277

## FHIR Team Checklist
- **Update the title** of the PR to be succinct and less than 65 characters
- **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- Tag the PR with the type of update: **Bug**, **Build**, **Dependencies**, **Enhancement**, **New-Feature** or **Documentation**
- Tag the PR with **Open source**, **Azure API for FHIR** (CosmosDB or common code) or **Azure Healthcare APIs** (SQL or common code) to specify where this change is intended to be released.
- Tag the PR with **Schema Version backward compatible** or **Schema Version backward incompatible** or **Schema Version unchanged** if this adds or updates Sql script which is/is not backward compatible with the code.
- [ ] CI is green before merge [Build Status](https://microsofthealthoss.visualstudio.com/FhirServer/_build/latest?definitionId=27&branchName=main) 
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Delete/DeletionService.cs [hunk 1]
```diff
@@ -192,7 +193,24 @@ public async Task<long> DeleteMultipleAsync(ConditionalDeleteResourceRequest req
...
+            {
+                // We need to wait until all running tasks are cancelled to get a count of resources deleted.
+                Task.WaitAll(deleteTasks.ToArray(), cancellationToken);
+            }
+            catch (AggregateException age)
+            {
+                // If one of the tasks fails, the rest may throw a cancellation exception. Filtering those out as they are noise.
+                foreach (var coreException in age.InnerExceptions.Where(e => e is not TaskCanceledException))
+                {
+                    _logger.LogError(age, "Error deleting");
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Delete/DeletionService.cs:206]: This is logging the aggregate exception multiple times. I think you want to log coreException.

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Delete/DeletionService.cs [hunk 2]
```diff
@@ -192,7 +193,24 @@ public async Task<long> DeleteMultipleAsync(ConditionalDeleteResourceRequest req
...
-            System.Threading.Tasks.Task.WaitAll(deleteTasks.ToArray(), cancellationToken);
+            try
+            {
+                // We need to wait until all running tasks are cancelled to get a count of resources deleted.
+                Task.WaitAll(deleteTasks.ToArray(), cancellationToken);
+            }
+            catch (AggregateException age)
+            {
+                // If one of the tasks fails, the rest may throw a cancellation exception. Filtering those out as they are noise.
+                foreach (var coreException in age.InnerExceptions.Where(e => e is not TaskCanceledException))
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Delete/DeletionService.cs:204]: Should TaskCancelledExceptions also be filtered out on line 223 when we make the exception to add to the job record?

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Delete/DeletionService.cs (hunk 1 above)
Comment (mikaelweave) [reply src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Delete/DeletionService.cs:206]: wow thank you need coffee

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Delete/DeletionService.cs (hunk 2 above)
Comment (mikaelweave) [reply src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Delete/DeletionService.cs:204]: I think we still want 223 (now 224) to count resources deleted on threads that have been cancelled??

I filtered out the `TaskCancelled` exceptions from the thrown `AggregateException`. Thoughts about this?

Comment (LTA-Thinking) [reply src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Delete/DeletionService.cs:204]: Yes, you what I was thinking of but didn't explain correctly.

**Review by LTA-Thinking — APPROVED** [review]

**Pull request merged**